### ISSUE TITLE
More encoding / decoding examples

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-common#11]
See https://github.com/ably/ably-ios/issues/95, I think client library developers will generally struggle to understand the spec and implement exactly as we expect.

As such, I recommend we introduce a new JSON file in https://github.com/ably/ably-common/tree/master/test-resources which includes all permutations of non-encrypted encodings & decodings to remove any ambiguity about what is expected.

@paddybyers @SimonWoolf WDYT?  This should not change the spec, but I can add a note in the encoding & decoding that is helps to run the tests against the JSON file.

**Comment by paddybyers** [comment]
> which includes all permutations of non-encrypted encodings & decodings to remove any ambiguity about what is expected.

Yes, but really we want examples that map a `Message`, or at least a `data`, to an encoded `Message` or `ProtocolMessage`, so it is obvious how the arguments presented in the API get turned into serialised entities. But .... we can't put binary data into a JSON file, so we need the file to include an extra level of encoding to be able to contain the `data` and the serialised entity. That could just add even more confusion unless we explain it carefully ...

**Comment by SimonWoolf** [comment]
:+1: good idea. there's a few in ably-js [here](https://github.com/ably/ably-js/blob/master/spec/realtime/message.test.js#L228-L304), but yeah, putting it all with expected decoded result in a common file is a good idea

**Comment by mattheworiordan** [comment]
> Yes, but really we want examples that map a Message, or at least a data, to an encoded Message or ProtocolMessage

I don't see why a `ProtocolMessage` is relevant given it will contain an array of `Message` objects, and encoding & decoding is contained within the `Message`

> But .... we can't put binary data into a JSON file, so we need the file to include an extra level of encoding to be able to contain the data and the serialised entity. That could just add even more confusion unless we explain it carefully ...

How is that any different to https://github.com/ably/ably-common/blob/master/test-resources/crypto-data-256.json#L9-L18?
